### PR TITLE
EHN add Haar-like feature

### PR DESCRIPTION
**Pull request by glemaitre**
## Description
[Tell us about your new feature, improvement, or fix! If relevant, please supplement the PR with images.]

## Checklist
[It's fine to submit PRs which are a work in progress! But before they are merged, all PRs should provide:]
- [x] Clean style in [the spirit of PEP8](https://www.python.org/dev/peps/pep-0008/)
- [x] [Docstrings for all functions](https://github.com/numpy/numpy/blob/master/doc/example.py)
- [x] Gallery example in `./doc/examples` (new features only)
- [x] Unit tests

[For detailed information on these and other aspects see [scikit-image contribution guidelines](http://scikit-image.org/docs/dev/contribute.html)]

## References
[If this is a bug-fix or enhancement, it closes issue # ]
[If this is a new feature, it implements the following paper: ]

## For reviewers

(Don't remove the checklist below.)

- [ ] Check that the PR title is short, concise, and will make sense 1 year
  later.
- [ ] Check that new functions are imported in corresponding `__init__.py`.
- [ ] Check that new features, API changes, and deprecations are mentioned in
      `doc/release/release_dev.rst`.

### CODE REVIEW

**Comment by glemaitre** [comment]
ping @warmspringwinds @stefanv @emmanuelle Here comes the core of the Haar-like features.

**Comment by warmspringwinds** [comment]
@glemaitre Great, thank you so much for your changes! :)

Just one more suggestion:
Can we put a part of your gallery example with feature visualization here (not with feature selection)?
I think it just belongs more here and will be similar to what we have with [local binary patterns](http://scikit-image.org/docs/dev/auto_examples/features_detection/plot_multiblock_local_binary_pattern.html).

But still leave the PR with feature selection (but without feature visualization), you can just reference
the corresponding page later.

After that make sure that all checks pass and we are good :+1:

File: skimage/feature/_haar.pyx [hunk 1]
```diff
@@ -0,0 +1,314 @@
...
+
+
+cpdef haar_like_feature_wrapper(
+    cnp.ndarray[integral_floating, ndim=2] int_image,
+    r, c, width, height, feature_type, feature_coord):
+    """Compute the Haar-like features for a region of interest (ROI) of an
+    integral image.
+
+    Haar-like features have been successively used in different computer vision
+    applications to detect different targets, objects, etc. It was first
```
Comment (warmspringwinds) [inline skimage/feature/_haar.pyx]: have been successively used for image classification and object detection [1]_

File: skimage/feature/haar.py [hunk 2]
```diff
@@ -0,0 +1,313 @@
...
+    return np.concatenate(feat_coord), np.hstack(feat_type)
+
+
+def haar_like_feature(int_image, r, c, width, height, feature_type=None,
+                      feature_coord=None):
+    """Compute the Haar-like features for a region of interest (ROI) of an
+    integral image.
+
+    Haar-like features have been successively used in different computer vision
+    applications to detect different targets, objects, etc. It was first
```
Comment (warmspringwinds) [inline skimage/feature/haar.py]: have been successively used for image classification and object detection [1]_

File: skimage/feature/haar.py [hunk 3]
```diff
@@ -0,0 +1,313 @@
...
+
+        - 'type-2-x': 2 rectangles varying along the x axis;
+        - 'type-2-y': 2 rectangles varying along the y axis;
+        - 'type-3-x': 3 rectangles varying along the x axis;
+        - 'type-3-y': 3 rectangles varying along the y axis;
+        - 'type-4': 4 rectangles varying along x and y axis.
+
+        By default all features are extracted.
+
+        If using with `feature_coord`, it should corresponds to the feature
```
Comment (warmspringwinds) [inline skimage/feature/haar.py]: should correspond

File: skimage/feature/haar.py [hunk 4]
```diff
@@ -0,0 +1,313 @@
...
+        type of each associated coordinate feature.
```
Comment (warmspringwinds) [inline skimage/feature/haar.py]: it should corresponds to the type of each associated feature.

File: skimage/feature/haar.py [hunk 5]
```diff
@@ -0,0 +1,313 @@
...
+        haar_feature[haar_feature_idx] = haar_feature.copy()
+        return haar_feature
+
+
+def draw_haar_like_feature(image, r, c, width, height,
+                           feature_coord,
+                           color_positive_block=(1., 0., 0.),
+                           color_negative_block=(0., 1., 0.),
+                           alpha=0.5, max_n_features=None, random_state=None):
+    """Helper to visualize Haar-like features.
```
Comment (warmspringwinds) [inline skimage/feature/haar.py]: Helper function for visualization of Haar-like features.

File: skimage/feature/haar.py (hunk 4 above)
Comment (warmspringwinds) [reply skimage/feature/haar.py]: it should correspond<del>s</del> to the type of each associated feature.

**Comment by glemaitre** [comment]
> Can we put a part of your gallery example with feature visualization here (not with feature selection)?
I think it just belongs more here and will be similar to what we have with local binary patterns.

I will do that later tonight or tomorrow. I think that this is possible.

**Comment by glemaitre** [comment]
@warmspringwinds I would suggest to take the following example: https://github.com/scikit-image/scikit-image/pull/2898/files up to line 70

**Comment by warmspringwinds** [comment]
@glemaitre Yes, up to line 70 would be great!
Once again -- great work.

File: skimage/feature/_haar.pyx [hunk 6]
```diff
@@ -0,0 +1,313 @@
+#cython: cdivision=True
+#cython: boundscheck=False
+#cython: nonecheck=False
+#cython: wraparound=False
+#distutils: language=c++
+
+from libc.stdlib cimport malloc, free, realloc
```
Comment (warmspringwinds) [inline skimage/feature/_haar.pyx]: dead imports?

**Comment by glemaitre** [comment]
Ok done. @warmspringwinds shall I move the feature selection example inside the `xx_applications` folder?

**Comment by emmanuelle** [comment]
I'd rather have a new section on image processing and machine learning. What do you think @scikit-image/core ?

**Comment by soupault** [comment]
From my perspective, `xx_applications` is an appropriate place for now. We can branch a dedicated section once we have a critical number of CV+ML/DL examples.

**Comment by stefanv** [comment]
Image processing + ML is quite a wide topic, so I'm not sure how these will be grouped together in the future; shall we wait until we have a few more examples to see how they fall into place?

File: doc/examples/features_detection/plot_haar.py [hunk 7]
```diff
@@ -0,0 +1,70 @@
+"""
+============================
+Haar-like feature descriptor
+============================
+
+Haar-like feature descriptors are a set of simple features which was used in
```
Comment (warmspringwinds) [inline doc/examples/features_detection/plot_haar.py]: Haar-like features are simple digital image features that were used in the first ...

File: doc/examples/features_detection/plot_haar.py [hunk 8]
```diff
@@ -0,0 +1,70 @@
...
+the first real-time face detector [1]_. These features can be efficiently
+computed on any scale in constant time, using an integral image [1]_ and by
```
Comment (warmspringwinds) [inline doc/examples/features_detection/plot_haar.py]: make it simpler:
 These features can be efficiently
 computed on any scale in constant time, using an integral image [1]_.

File: doc/examples/features_detection/plot_haar.py [hunk 9]
```diff
@@ -0,0 +1,70 @@
...
+from skimage.feature import haar_like_feature_coord
+from skimage.feature import draw_haar_like_feature
+
+print(__doc__)
+
+###############################################################################
+# The different Haar-like feature descriptors
+###############################################################################
+# The Haar-like feature descriptors come into 5 different types as illustrated
+# in the figure below. The value of the descriptor consists in the subtraction
```
Comment (warmspringwinds) [inline doc/examples/features_detection/plot_haar.py]: The value of the descriptor is equal to the difference of sums of
intensity values between the green and red rectangles.

File: doc/examples/features_detection/plot_haar.py [hunk 10]
```diff
@@ -0,0 +1,70 @@
...
+    ax.set_title(feat_t)
+    ax.set_xticks([])
+    ax.set_yticks([])
+
+fig.suptitle('The different Haar-like feature descriptors')
+plt.axis('off')
+plt.show()
+
+###############################################################################
+# The feature scalar value corresponds to the sum of the pixel intensities of
```
Comment (warmspringwinds) [inline doc/examples/features_detection/plot_haar.py]: is equal to the sum of pixel intensities

File: doc/examples/features_detection/plot_haar.py [hunk 11]
```diff
@@ -0,0 +1,70 @@
...
+# the red area is subtracted to the sum of the pixel intensities of the green
+# area. In practice, the haar-like features will be placed in all possible
+# location of an image and a feature value will be extracted for each of these
```
Comment (warmspringwinds) [inline doc/examples/features_detection/plot_haar.py]: a feature value will be computed

File: doc/examples/features_detection/plot_haar.py [hunk 12]
```diff
@@ -0,0 +1,70 @@
...
+import numpy as np
+import matplotlib.pyplot as plt
+
+from skimage.feature import haar_like_feature_coord
+from skimage.feature import draw_haar_like_feature
+
+print(__doc__)
+
+###############################################################################
+# The different Haar-like feature descriptors
```
Comment (warmspringwinds) [inline doc/examples/features_detection/plot_haar.py]: Different types of Haar-like feature descriptor

File: skimage/feature/_haar.pyx [hunk 13]
```diff
@@ -0,0 +1,311 @@
...
+
+
+cpdef haar_like_feature_wrapper(
+    cnp.ndarray[integral_floating, ndim=2] int_image,
+    r, c, width, height, feature_type, feature_coord):
+    """Compute the Haar-like features for a region of interest (ROI) of an
+    integral image.
+
+    Haar-like features have been successively used for image classification and
+    object detection [1]_. It has been widely used for real-time face detection
```
Comment (warmspringwinds) [inline skimage/feature/_haar.pyx]: It has been <del>widely</del> used for real-time face detection

File: skimage/feature/_haar.pyx [hunk 14]
```diff
@@ -0,0 +1,311 @@
...
+        - 'type-2-x': 2 rectangles varying along the x axis;
+        - 'type-2-y': 2 rectangles varying along the y axis;
+        - 'type-3-x': 3 rectangles varying along the x axis;
+        - 'type-3-y': 3 rectangles varying along the y axis;
+        - 'type-4': 4 rectangles varying along x and y axis.
+
+    Returns
+    -------
+    haar_features : (n_features,) ndarray
+        Resulting Haar-like features. Each value corresponds to the subtraction
```
Comment (warmspringwinds) [inline skimage/feature/_haar.pyx]: subtraction of intensity sums...

File: skimage/feature/haar.py [hunk 15]
```diff
@@ -0,0 +1,312 @@
...
+        fully transparent, 0 - opaque.
+    max_n_features : int, default=None
+        The maximum number of features to be returned.
+        By default, all features are returned.
+    random_state : int, RandomState instance or None, optional
+        If int, random_state is the seed used by the random number generator;
+        If RandomState instance, random_state is the random number generator;
+        If None, the random number generator is the RandomState instance used
+        by `np.random`. The random state is used when generating a set of
+        feature smaller than the total number of available features.
```
Comment (warmspringwinds) [inline skimage/feature/haar.py]: a set of features smaller than ...

**Comment by warmspringwinds** [comment]
@glemaitre awesome work.

I will generate docs and see how the example look like and get back to you.

**Comment by glemaitre** [comment]
@warmspringwinds @stefanv any follow up?

File: skimage/feature/haar.py [hunk 16]
```diff
@@ -0,0 +1,312 @@
...
+
+    return np.concatenate(feat_coord), np.hstack(feat_type)
+
+
+def haar_like_feature(int_image, r, c, width, height, feature_type=None,
+                      feature_coord=None):
+    """Compute the Haar-like features for a region of interest (ROI) of an
+    integral image.
+
+    Haar-like features have been successively used for image classification and
```
Comment (warmspringwinds) [inline skimage/feature/haar.py]: successfully

File: skimage/feature/haar.py [hunk 17]
```diff
@@ -0,0 +1,312 @@
...
+        haar_feature[haar_feature_idx] = haar_feature.copy()
+        return haar_feature
+
+
+def draw_haar_like_feature(image, r, c, width, height,
+                           feature_coord,
+                           color_positive_block=(1., 0., 0.),
+                           color_negative_block=(0., 1., 0.),
+                           alpha=0.5, max_n_features=None, random_state=None):
+    """Helper function for visualization Haar-like features.
```
Comment (warmspringwinds) [inline skimage/feature/haar.py]: visualization of

File: doc/examples/features_detection/plot_haar.py [hunk 18]
```diff
@@ -0,0 +1,69 @@
+"""
+============================
+Haar-like feature descriptor
+============================
+
+Haar-like features are simple digital image features that were used first in a
```
Comment (warmspringwinds) [inline doc/examples/features_detection/plot_haar.py]: were introduced in a ...

File: doc/examples/features_detection/plot_haar.py [hunk 19]
```diff
@@ -0,0 +1,69 @@
...
+fig.suptitle('The different Haar-like feature descriptors')
+plt.axis('off')
+plt.show()
+
+###############################################################################
+# The value of the descriptor is equal to the difference between the sum of the
+# intensity values in the green rectangle and the red one.  the red area is
+# subtracted to the sum of the pixel intensities of the green In practice, the
+# Haar-like features will be placed in all possible location of an image and a
+# feature value will be computed for each of these locations.
```
Comment (warmspringwinds) [inline doc/examples/features_detection/plot_haar.py]: After that, a small number of critical features is selected from this large set of
potential features (for example, using AdaBoost learning algorithm as in [1]). -- just add this for clarity

**Comment by glemaitre** [comment]
Done!

**Comment by warmspringwinds** [comment]
@glemaitre Thank you :+1: 

@stefanv I think we are ready :)

**Comment by stefanv** [comment]
Excellent work @glemaitre ; thank you!

**Pull request merged**